diff --git a/Edk2/Protocol/DiskIo2.h b/Edk2/Protocol/DiskIo2.h
index 6e0665b..72d978e 100644
--- a/Edk2/Protocol/DiskIo2.h
+++ b/Edk2/Protocol/DiskIo2.h
@@ -20,7 +20,7 @@
 
 #define EFI_DISK_IO2_PROTOCOL_GUID \
   { \
-    0x151c8eae, 0x7f2c, 0x472c, 0x9e, 0x54, 0x98, 0x28, 0x19, 0x4f, 0x6a, 0x88 \
+    0x151c8eae, 0x7f2c, 0x472c, {0x9e, 0x54, 0x98, 0x28, 0x19, 0x4f, 0x6a, 0x88} \
   }
 
 typedef struct _EFI_DISK_IO2_PROTOCOL EFI_DISK_IO2_PROTOCOL;
