--- a/clloader.c	2014-08-18 16:16:43.904000000 -0700
+++ b/clloader.c	2014-08-22 14:29:57.904000000 -0700
@@ -2721,7 +2721,13 @@
 						ret = close(mystate.tty_from_host);
 						fprintf(stderr, "Bad connection to host  %d bytes available, ioctl return is %d\n",rd_bytes, ioctl_rc) ;
 						mystate.tty_from_host = -1 ;
+						fprintf(stderr, "cable was probably disconnected\n");
+						//kill sketch then kill clloader
+						kill(mystate.slave_pid, SIGTERM);
+						exit(0);
 					}else{
+						fprintf(stderr, "host: %s\n", rbuf_from_host);
+						errors = 0;
 						/* TODO: bring out state change of BAUD/LINE and replace magic string */
 						if (strncmp(CL_LOADER_CMD_HOST_START_DOWNLOAD_CMD,
 							rbuf_from_host,
