--- a/clloader.h	2015-01-06 10:48:26.000000000 -0800
+++ b/clloader.h	2015-01-06 10:51:46.000000000 -0800
@@ -28,12 +28,13 @@
 
 
 #define _XOPEN_SOURCE       /* See feature_test_macros(7) */
 #define MAX_ARGS 0x20
 
 #define CL_LOADER_CMD_HOST_START_DOWNLOAD_CMD "~sketch download"
+#define downloadCMD "~sketch downloadGalileo"
 
 #define CL_LOADER_CMD_STARTCHAR '#'
 #define CL_LOADER_CMD_START_SKETCH_APP "#Start Sketch:" /* filename args and exec i/o wrapper */
 
 extern int clLoaderFSM(int argc, char **argp);
 
