From a21235f0c837b78bfea9b463a01e3b1efe613195 Mon Sep 17 00:00:00 2001
From: "dunno@dunno" <dunno@dunno>
Date: Thu, 25 Oct 2018 01:28:05 -0400
Subject: [PATCH 2/2] Pk/CryptTs.c fix SetMem

---
 Cryptlib/Pk/CryptTs.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Cryptlib/Pk/CryptTs.c b/Cryptlib/Pk/CryptTs.c
index d63c23d..644a782 100644
--- a/Cryptlib/Pk/CryptTs.c
+++ b/Cryptlib/Pk/CryptTs.c
@@ -161,7 +161,7 @@ ConvertAsn1TimeToEfiTime (
   }
 
   Str = (CONST CHAR8*)Asn1Time->data;
-  SetMem (EfiTime, 0, sizeof (EFI_TIME));
+  SetMem (EfiTime, sizeof (EFI_TIME), 0);
 
   Index = 0;
   if (Asn1Time->type == V_ASN1_UTCTIME) {               /* two digit year */
-- 
2.14.1

