diff --git i/init-live.sh w/init-live.sh
index 4bd1b52..cfc57b4 100644
--- i/init-live.sh
+++ w/init-live.sh
@@ -3,7 +3,7 @@
 PATH=/sbin:/bin:/usr/sbin:/usr/bin
 
 ROOT_MOUNT="/rootfs/"
-ROOT_IMAGE="rootfs.img"
+ROOT_IMAGE="image-full-clanton.ext3"
 MOUNT="/bin/mount"
 UMOUNT="/bin/umount"
 ISOLINUX=""
@@ -50,6 +50,8 @@ read_args() {
         case $arg in
             root=*)
                 ROOT_DEVICE=$optarg ;;
+            rootimage=*)
+                ROOT_IMAGE=$optarg ;;
             rootfstype=*)
                 modprobe $optarg 2> /dev/null ;;
             LABEL=*)
@@ -79,14 +81,19 @@ boot_live_root() {
 
     # Move the mount points of some filesystems over to
     # the corresponding directories under the real root filesystem.
-    mount -n --move /proc ${ROOT_MOUNT}/proc
-    mount -n --move /sys ${ROOT_MOUNT}/sys
-    mount -n --move /dev ${ROOT_MOUNT}/dev
+    mount -n --move /proc ${ROOT_MOUNT}/proc ||
+      fatal "Couldn't remount /proc, dropping to shell"
+    mount -n --move /sys ${ROOT_MOUNT}/sys ||
+      fatal "Couldn't remount /sys, dropping to shell"
+    mount -n --move /dev ${ROOT_MOUNT}/dev ||
+      fatal "Couldn't remount /dev, dropping to shell"
     # Move /media/$i over to the real root filesystem
-    mount -n --move /media/$i ${ROOT_MOUNT}/media/realroot
+    mount -n --move /media/$i ${ROOT_MOUNT}/media/realroot ||
+      fatal "Couldn't remount /media/$i, dropping to shell"
 
     cd $ROOT_MOUNT
-    exec switch_root -c /dev/console $ROOT_MOUNT /sbin/init
+    exec switch_root -c /dev/console $ROOT_MOUNT /sbin/init ||
+      fatal "Couldn't switch_root, dropping to shell"
 }
 
 fatal() {
@@ -127,7 +134,7 @@ do
            mount | grep media
            echo "Available block devices"
            ls /dev/sd*
-           fatal "Cannot find rootfs.img file in /media/* , dropping to a shell "
+           fatal "Cannot find $ROOT_IMAGE file in /media/* , dropping to a shell "
       fi
       C=$(( C + 1 ))
   fi
